fastcrypto-vdf/src/class_group/hash.rs

```diff
@@ -23,105 +22,90 @@ use crate::math::jacobi;
-            .expect("a and b are constructed such that this never fails"))
-    }
-
-    /// Generate a random quadratic form from a seed with the given discriminant. This method is deterministic, and it
-    /// is a random oracle on a large subset of the class group. This method picks a default `k` parameter and calls the
-    /// [hash_to_group](QuadraticForm::hash_to_group) function with this `k`.
-    ///
-    /// This method returns an InvalidInput error if the discriminant is smaller than 800 bits.
-    pub fn hash_to_group_with_default_parameters(
-        seed: &[u8],
-        discriminant: &Discriminant,
-    ) -> FastCryptoResult<Self> {
-        let k = get_default_k(discriminant.bits() as usize);
-        Self::hash_to_group(seed, discriminant, k)
+        // Parameters are chosen such that the image of the hash function is ~260 bits to avoid collision and, and each prime factor is ~130 bits to avoid them being precomputed.
+        hash_to_group_with_custom_parameters(seed, discriminant, 17, 2)
     }
 }
 
-fn get_default_k(discriminant_bits: usize) -> u16 {
-    // This is chosen to ensure that the range of the hash function is large (at least 2^256) but also that the
-    // performance is near optimal, based on benchmarks.
-    if discriminant_bits <= 2048 {
-        16
-    } else {
-        32
+/// Generate a random quadratic form from a seed with the given discriminant and custom parameters.
+///
+/// The output will be a uniformly random element from the set of points (a,b,c) where a = p_1 ... p_k
+/// for some primes p_i < 2^lambda.
+///
+/// If the discriminant is not a negative prime, an [InvalidInput] error may be returned.
+///
+/// The parameters must be chosen carefully to ensure that the function is secure and for all
+/// use cases, [hash_to_group] should be used.
+fn hash_to_group_with_custom_parameters(
+    seed: &[u8],
+    discriminant: &Discriminant,
+    prime_factor_size_in_bytes: u64,
+    prime_factors: u64,
+) -> FastCryptoResult<QuadraticForm> {
+    // Ensure that the image is sufficiently large
+    debug_assert!(prime_factors as f64 * n_bit_primes(prime_factor_size_in_bytes * 8) >= 256f64);
+
+    // Ensure that the prime factors are so large that the corresponding quadratic form be precomputed.
+    debug_assert!(n_bit_primes(prime_factor_size_in_bytes * 8) >= 128f64);
+
+    // Ensure that the result will be reduced
+    debug_assert!(
+        discriminant.as_bigint().abs().sqrt().shr(1)
+            > BigInt::one().shl(prime_factors * prime_factor_size_in_bytes)
+    );
+
+    // Sample a and b such that a < sqrt(|discriminant|)/2 has exactly prime_factors prime factors and b is the square root of the discriminant modulo a.
+    let (a, mut b) = sample_modulus(
+        seed,
+        discriminant,
+        prime_factor_size_in_bytes,
+        prime_factors,
+    )?;
+
+    // b must be odd but may be negative
+    if b.is_even() {
+        b -= &a;
```

Comment:
what does this offer?

Reply:
This picks the right square root to ensure that the quadratic form is valid and reduced.